### PR TITLE
backend/gcs: Stop using legacy helper/schema

### PR DESCRIPTION
**Pull request by apparentlymart**
As part of our efforts to remove the large snapshot of the legacy SDK that we've been using, this replaces the uses of that package with our new lightweight `backendbase` package, which is a collection of helpers that can substitute for most of the parts of the legacy SDK that backends tend to use.

For this backend for now we'll use the "SDK-like" helpers which aim to retain some of the legacy-SDK-specific assumptions that aren't technically true anymore, such as the idea that null and empty string are equivalent. Hopefully one day we'll be able to wean this backend off of that more completely, but this is a pragmatic way to get away from the legacy SDK without a large rewrite.

This also drew attention to the fact that this backend was previously relying on the weird context.Context value that the SDK was passing in to the configure method, even though that context isn't connected up to anything particularly useful. In the long run we should change the backend APIs to pass context.Context to each method that might make network requests, but for now we're using `context.TODO` so that we can more easily find these cases and update them later once there's a real context to use.

In the long run we're still planning to move all of the state storage backends out to provider plugins, but the legacy SDK is quite a liability because few people know how to maintain it while hopefully this interim `backendbase` package is easier to maintain (if needed) due to its relative simplicity.

I don't have any test environment for this backend, so I've only tested this through its unit tests. I'd appreciate if a backend maintainer could, along with reviewing this code, also run the acceptance test suite in case there are regressions that the unit tests aren't able to catch. Thanks!

### CODE REVIEW

**Comment by SarahFrench** [comment]
I've taken a quick look and the code changes make sense to me, but tomorrow I'll run the acceptance tests and either do manual tests or identify additional acceptance tests we need to ensure the "SDK-like" helpers are doing their job.

When we muxed the Google provider we realised that there were no tests asserting how empty strings should be handled, so we accidentally broke how some users had configured the provider (e.g. setting `credentials = ""`). We learned that often there can be a lot of behaviour you take as a given and don't think you need to test, until you really do 😂 

I'm excited about the new backendbase package. Previously I wanted to add some test helper functions to the backend folder to help with writing [tests similar to this](https://github.com/hashicorp/terraform-provider-google/blob/e3453f593143c495b265eb04ed2e452669f81c0c/google/provider/provider_internal_test.go#L168) for backends, but the old SDK was pretty intimidating/it looked like a lot of unpicking would be required to aid unit testing. It may be that now testing functions are not needed, or are a lot easier to introduce as an external contributor.

Like I said, I'll do some manual testing tomorrow. If there are any deadlines related to this PR please let me know!

**Review by SarahFrench — APPROVED** [review]
I've run the old acc tests locally, tested this manually, and did some testing via new acceptance tests I'm preparing in a separate branch.

My main concern was empty strings suddenly becoming un-ignored, but I can't see evidence of that occurring in [acc tests](https://github.com/hashicorp/terraform/pull/35026/commits/8d327adfd53648b9deca0dcb223d205e4e816166) or manual tests.

One difference I noticed this PR introduces is that if the gcs backend is configured with both kms_encryption_key and encryption_key the validation about them conflicting doesn't kick in.

```
terraform {
  backend "gcs" {
    bucket             = "my-bucket"
    kms_encryption_key = ""
    encryption_key = ""
  }
}
```

Previously `terraform init -reconfigure` resulted in:

```
Initializing the backend...
╷
│ Error: "encryption_key": conflicts with kms_encryption_key
│ 
│ 
╵

╷
│ Error: "kms_encryption_key": conflicts with encryption_key
│ 
│ 
```

but after this PR the conflict isn't raised.

I'm not majorly concerned because there's still an error if those fields are both set to non-trivial values, so _meaningful_ validation isn't broken and you still get:

```
Initializing the backend...
╷
│ Error: can't set both encryption_key and kms_encryption_key
│ 
│ 
╵
```

----

I'm about to be OOO until Wednesday next week so I'll approve this, but if the difference above requires more discussion and can't wait until Wednesday please ping @/hashicorp/tf-eco-hybrid-cloud

**Comment by apparentlymart** [comment]
Hmmm indeed that seems to be a result of implementing that validation in terms of the "SDK-like" helper's results, instead of inside the SDK itself, because the helper is designed to behave like `schema.ResourceData` in how it treats absent arguments whereas the SDK's conflict validation happens earlier and does not itself rely on the `ResourceData` abstraction.

I think we could potentially restore that precision by having that check in particular bypass the "SDK-like helper" and go directly to the underlying cty API.

Since this won't go out until Terraform v1.9 anyway and since you mentioned that you are not super concerned with this current compromise I'm going to merge this now just to get this baseline in place and then I'll try to work on a separate PR to restore the previous precision of this check which we can merge before the v1.9 release if it turns out to be doable.

Thanks for reviewing and testing!

**Pull request merged**
